modules/ppcp-button/src/Endpoint/ApproveOrderEndpoint.php

```
@@ -271,10 +268,79 @@ public function handle_request(): bool {
+	 * - 3DS details with a clear rejected: Payment fails.
+	 *
+	 * @param Order $order The PayPal order to inspect.
+	 * @throws RuntimeException When the 3DS check was rejected.
+	 */
+	protected function verify_three_d_secure( Order $order ) : void {
+		$payment_source = $order->payment_source();
+
+		if ( ! $payment_source ) {
+			// Missing 3DS details.
+			return;
+		}
+
+		$proceed      = ThreeDSecure::NO_DECISION;
+		$order_status = $order->status();
+		$source_name  = $payment_source->name();
+
+		/**
+		 * For GooglePay (and possibly other payment sources) we check the order
+		 * status, as it will clearly indicate if verification is needed.
+		 *
+		 * Note: PayPal is currently investigating this case.
+		 * Maybe the order status is wrong and should be ACCEPTED, in that case,
+		 * we could drop the condition and always run proceed_with_order().
+		 */
+		if ( $order_status->is( OrderStatus::PAYER_ACTION_REQUIRED ) ) {
+			$proceed = $this->threed_secure->proceed_with_order( $order );
+		} elseif ( 'card' === $source_name ) {
+			// For credit cards, we also check the 3DS response.
+			$proceed = $this->threed_secure->proceed_with_order( $order );
+		}
+
+		// Handle the verification result based on the proceed value.
+		switch ( $proceed ) {
+			case ThreeDSecure::PROCEED:
+				// Check was successful.
+				return;
+
+			case ThreeDSecure::NO_DECISION:
+				// No rejection. Let's proceed with the payment.
+				return;
+
+			case ThreeDSecure::RETRY:
+				// Rejection case 1, verification can be retried.
+				throw new RuntimeException(
+					__(
+						'Something went wrong. Please try again.',
+						'woocommerce-paypal-payments'
+					)
+				);
+
+			case ThreeDSecure::REJECT:
+				// Rejection case 2, payment was rejected.
+				throw new RuntimeException(
+					__(
+						'Unfortunately, we can\'t accept your card. Please choose a different payment method.',
+						'woocommerce-paypal-payments'
+					)
+				);
+		}
```

Are we certain that $proceed will always be one of the expected ThreeDSecure::* constants? If not, we should add a default case to handle unexpected values and avoid silent failures.

Also, since both `PROCEED` and `NO_DECISION` lead to the same outcome (return), we can group them to simplify the switch.


```
switch ( $proceed ) {
	case ThreeDSecure::PROCEED:
	case ThreeDSecure::NO_DECISION:
		// Payment can proceed.
		return;

	case ThreeDSecure::RETRY:
		// Rejection case 1, verification can be retried.
		throw new RuntimeException(
			__(
				'Something went wrong. Please try again.',
				'woocommerce-paypal-payments'
			)
		);

	case ThreeDSecure::REJECT:
		// Rejection case 2, payment was rejected.
		throw new RuntimeException(
			__(
				'Unfortunately, we can\'t accept your card. Please choose a different payment method.',
				'woocommerce-paypal-payments'
			)
		);

	default:
		// Failsafe: unknown status
		throw new RuntimeException(
			__(
				'Unexpected 3DS decision. Please try again or contact support.',
				'woocommerce-paypal-payments'
			)
		);
}

```